Class not found when unmarshalling: com.braintreepayments.api.dropin.DropInRequest

<!-- Only open an issue here if you think you've found an issue with our SDK. If you need help troubleshooting your integration, reach out to Braintree Support at https://help.braintreepayments.com. -->

### General information

* SDK/Library version: 4.6.0
* Environment: Sandbox
* Android Version and Device: Samsung Galaxy S10 with Android 10.0

### Issue description

<!-- To help us quickly reproduce your issue, include as many details as possible, such as logs, steps to reproduce, and so on.  If the issue reports a new feature, follow the [user story](https://en.wikipedia.org/wiki/User_story) format to clearly describe the use case. -->

I cannot even make work the "hello world" equivalent described on this page: https://developers.braintreepayments.com/start/hello-client/android/v3

I have added the following snippet:

```
        button.setOnClickListener {
            val dropInRequest = DropInRequest()
                    .clientToken("CLIENT_TOKEN_FROM_SERVER")
            startActivityForResult(dropInRequest.getIntent(this), REQUEST_CODE)
        }
```

And I get the following stack trace:

```
2020-08-18 10:32:09.777 5841-15936/? E/Parcel: Class not found when unmarshalling: com.braintreepayments.api.dropin.DropInRequest
    java.lang.ClassNotFoundException: com.braintreepayments.api.dropin.DropInRequest
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:454)
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031)
        at android.os.Parcel.readParcelable(Parcel.java:2981)
        at android.os.Parcel.readValue(Parcel.java:2883)
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261)
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292)
        at android.os.BaseBundle.unparcel(BaseBundle.java:236)
        at android.os.BaseBundle.getString(BaseBundle.java:1160)
        at android.content.Intent.getStringExtra(Intent.java:8548)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:790)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:694)
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:1925)
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:625)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1707)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1603)
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1557)
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1690)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
     Caused by: java.lang.ClassNotFoundException: com.braintreepayments.api.dropin.DropInRequest
        at java.lang.Class.classForName(Native Method) 
        at java.lang.Class.forName(Class.java:454) 
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031) 
        at android.os.Parcel.readParcelable(Parcel.java:2981) 
        at android.os.Parcel.readValue(Parcel.java:2883) 
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261) 
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292) 
        at android.os.BaseBundle.unparcel(BaseBundle.java:236) 
        at android.os.BaseBundle.getString(BaseBundle.java:1160) 
        at android.content.Intent.getStringExtra(Intent.java:8548) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:790) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:694) 
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:1925) 
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:625) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1707) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1603) 
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1557) 
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1690) 
        at android.os.Binder.execTransactInternal(Binder.java:1021) 
        at android.os.Binder.execTransact(Binder.java:994) 
```

Someone got the same error detailed here: https://stackoverflow.com/questions/54221312/braintree-integration-error-class-not-found-com-braintreepayments-api-dropin but no answer.

Please note that I have even tried with a token generated from my server, but this should not be taken into account.

Moreover I cannot understand the situation which has been discussed in #57.

So if you guys have an idea of what's going on, any help would be very appreciated.

Thanks!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,9 @@
   * Add `VenmoRequest` setter to `DropInRequest`
 * CardForm
   * Focus Cardholder Name field when adding card details if field not disabled (fixes #180)
+* DropInRequest
+  * Set explicit class loader for `DropInRequest` (fixes #191)
+  * Wrap `DropInRequest` in Parcelable bundle (fixes #228)
 * Breaking Changes
   * Visibility
     * Make `BottomSheetState` package-private

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
@@ -90,6 +90,7 @@ public class DropInActivity extends AppCompatActivity {
         }
 
         alertPresenter = new AlertPresenter();
+        clientTokenPresent = dropInClient.getAuthorization() instanceof ClientToken;
         dropInRequest = getDropInRequest(getIntent());
 
         dropInViewModel = new ViewModelProvider(this).get(DropInViewModel.class);

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
@@ -90,8 +90,8 @@ public class DropInActivity extends AppCompatActivity {
         }
 
         alertPresenter = new AlertPresenter();
-        clientTokenPresent = dropInClient.getAuthorization() instanceof ClientToken;
         dropInRequest = getDropInRequest(getIntent());
+        clientTokenPresent = dropInClient.getAuthorization() instanceof ClientToken;
 
         dropInViewModel = new ViewModelProvider(this).get(DropInViewModel.class);
         fragmentContainerView = findViewById(R.id.fragment_container_view);

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
@@ -81,7 +81,7 @@ public class DropInActivity extends AppCompatActivity {
             Intent intent = getIntent();
             String authorization = intent.getStringExtra(DropInClient.EXTRA_AUTHORIZATION);
             String sessionId = intent.getStringExtra(DropInClient.EXTRA_SESSION_ID);
-            DropInRequest dropInRequest = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST);
+            DropInRequest dropInRequest = getDropInRequest(intent);
             dropInClient = new DropInClient(this, authorization, sessionId, dropInRequest);
         }
 
@@ -92,7 +92,7 @@ public class DropInActivity extends AppCompatActivity {
         }
 
         alertPresenter = new AlertPresenter();
-        dropInRequest = getIntent().getParcelableExtra(EXTRA_CHECKOUT_REQUEST);
+        dropInRequest = getDropInRequest(getIntent());
 
         dropInViewModel = new ViewModelProvider(this).get(DropInViewModel.class);
         fragmentContainerView = findViewById(R.id.fragment_container_view);
@@ -149,6 +149,13 @@ public class DropInActivity extends AppCompatActivity {
         finish();
     }
 
+    private DropInRequest getDropInRequest(Intent intent) {
+        Bundle bundle = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST);
+        bundle.setClassLoader(DropInRequest.class.getClassLoader());
+        return bundle.getParcelable(DropInClient.EXTRA_CHECKOUT_REQUEST);
+    }
+
+    @VisibleForTesting
     void onDropInEvent(DropInEvent event) {
         switch (event.getType()) {
             case ADD_CARD_SUBMIT:

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
@@ -1,7 +1,5 @@
 package com.braintreepayments.api;
 
-import static com.braintreepayments.api.DropInClient.EXTRA_CHECKOUT_REQUEST;
-
 import android.content.Intent;
 import android.os.Bundle;
 
@@ -150,7 +148,7 @@ public class DropInActivity extends AppCompatActivity {
     }
 
     private DropInRequest getDropInRequest(Intent intent) {
-        Bundle bundle = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST);
+        Bundle bundle = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST_BUNDLE);
         bundle.setClassLoader(DropInRequest.class.getClassLoader());
         return bundle.getParcelable(DropInClient.EXTRA_CHECKOUT_REQUEST);
     }

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInActivity.java
@@ -1,5 +1,7 @@
 package com.braintreepayments.api;
 
+import static com.braintreepayments.api.DropInClient.EXTRA_CHECKOUT_REQUEST;
+
 import android.content.Intent;
 import android.os.Bundle;
 
@@ -90,8 +92,7 @@ public class DropInActivity extends AppCompatActivity {
         }
 
         alertPresenter = new AlertPresenter();
-        dropInRequest = getIntent().getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST);
-        clientTokenPresent = dropInClient.getAuthorization() instanceof ClientToken;
+        dropInRequest = getIntent().getParcelableExtra(EXTRA_CHECKOUT_REQUEST);
 
         dropInViewModel = new ViewModelProvider(this).get(DropInViewModel.class);
         fragmentContainerView = findViewById(R.id.fragment_container_view);

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
@@ -21,6 +21,7 @@ import java.util.Set;
 public class DropInClient {
 
     public static final String EXTRA_CHECKOUT_REQUEST = "com.braintreepayments.api.EXTRA_CHECKOUT_REQUEST";
+    public static final String EXTRA_CHECKOUT_REQUEST_BUNDLE = "com.braintreepayments.api.EXTRA_CHECKOUT_REQUEST_BUNDLE";
     public static final String EXTRA_SESSION_ID = "com.braintreepayments.api.EXTRA_SESSION_ID";
     public static final String EXTRA_AUTHORIZATION = "com.braintreepayments.api.EXTRA_AUTHORIZATION";
 
@@ -386,7 +387,7 @@ public class DropInClient {
 
     public void launchDropInForResult(FragmentActivity activity, int requestCode) {
         Bundle dropInRequestBundle = new Bundle();
-        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST, dropInRequest);
+        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST_BUNDLE, dropInRequest);
         Intent intent = new Intent(activity, DropInActivity.class)
                 .putExtra(EXTRA_CHECKOUT_REQUEST, dropInRequestBundle)
                 .putExtra(EXTRA_SESSION_ID, braintreeClient.getSessionId())

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
@@ -2,6 +2,7 @@ package com.braintreepayments.api;
 
 import android.content.Context;
 import android.content.Intent;
+import android.os.Bundle;
 import android.text.TextUtils;
 
 import androidx.annotation.Nullable;
@@ -384,8 +385,10 @@ public class DropInClient {
     }
 
     public void launchDropInForResult(FragmentActivity activity, int requestCode) {
+        Bundle dropInRequestBundle = new Bundle();
+        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST, dropInRequest);
         Intent intent = new Intent(activity, DropInActivity.class)
-                .putExtra(EXTRA_CHECKOUT_REQUEST, dropInRequest)
+                .putExtra(EXTRA_CHECKOUT_REQUEST, dropInRequestBundle)
                 .putExtra(EXTRA_SESSION_ID, braintreeClient.getSessionId())
                 .putExtra(EXTRA_AUTHORIZATION, braintreeClient.getAuthorization().toString());
         activity.startActivityForResult(intent, requestCode);

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java
@@ -387,9 +387,9 @@ public class DropInClient {
 
     public void launchDropInForResult(FragmentActivity activity, int requestCode) {
         Bundle dropInRequestBundle = new Bundle();
-        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST_BUNDLE, dropInRequest);
+        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST, dropInRequest);
         Intent intent = new Intent(activity, DropInActivity.class)
-                .putExtra(EXTRA_CHECKOUT_REQUEST, dropInRequestBundle)
+                .putExtra(EXTRA_CHECKOUT_REQUEST_BUNDLE, dropInRequestBundle)
                 .putExtra(EXTRA_SESSION_ID, braintreeClient.getSessionId())
                 .putExtra(EXTRA_AUTHORIZATION, braintreeClient.getAuthorization().toString());
         activity.startActivityForResult(intent, requestCode);

--- a/Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java
+++ b/Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java
@@ -1,11 +1,14 @@
 package com.braintreepayments.api;
 
+import android.os.Bundle;
 import android.os.Parcel;
 import android.os.Parcelable;
 
 import androidx.annotation.Nullable;
 
 import com.braintreepayments.cardform.view.CardForm;
+
+import java.io.Serializable;
 
 /**
  * Used to start {@link DropInActivity} with specified options.

--- a/Drop-In/src/test/java/com/braintreepayments/api/DropInActivityTest.kt
+++ b/Drop-In/src/test/java/com/braintreepayments/api/DropInActivityTest.kt
@@ -21,9 +21,6 @@ import com.braintreepayments.api.DropInClient.EXTRA_CHECKOUT_REQUEST
 
 import android.os.Bundle
 
-
-
-
 @RunWith(RobolectricTestRunner::class)
 class DropInActivityTest {
 

--- a/Drop-In/src/test/java/com/braintreepayments/api/DropInActivityTest.kt
+++ b/Drop-In/src/test/java/com/braintreepayments/api/DropInActivityTest.kt
@@ -17,6 +17,12 @@ import org.robolectric.RobolectricTestRunner
 import org.robolectric.Shadows.shadowOf
 import org.robolectric.android.controller.ActivityController
 import java.util.*
+import com.braintreepayments.api.DropInClient.EXTRA_CHECKOUT_REQUEST
+
+import android.os.Bundle
+
+
+
 
 @RunWith(RobolectricTestRunner::class)
 class DropInActivityTest {
@@ -1008,8 +1014,10 @@ class DropInActivityTest {
 
     private fun setupDropInActivity(dropInClient: DropInClient, dropInRequest:DropInRequest) {
         val context = InstrumentationRegistry.getInstrumentation().targetContext
+        val dropInRequestBundle = Bundle()
+        dropInRequestBundle.putParcelable(EXTRA_CHECKOUT_REQUEST, dropInRequest)
         val intent = Intent(context, DropInActivity::class.java)
-        intent.putExtra(DropInClient.EXTRA_CHECKOUT_REQUEST, dropInRequest)
+        intent.putExtra(DropInClient.EXTRA_CHECKOUT_REQUEST_BUNDLE, dropInRequestBundle)
 
         activityController = buildActivity(DropInActivity::class.java, intent)
         activity = activityController.get()

--- a/Drop-In/src/test/java/com/braintreepayments/api/DropInClientUnitTest.java
+++ b/Drop-In/src/test/java/com/braintreepayments/api/DropInClientUnitTest.java
@@ -16,6 +16,7 @@ import static org.mockito.Mockito.when;
 
 import android.content.Intent;
 import android.net.Uri;
+import android.os.Bundle;
 
 import androidx.fragment.app.FragmentActivity;
 
@@ -1404,7 +1405,9 @@ public class DropInClientUnitTest {
         assertEquals("session-id", intent.getStringExtra(DropInClient.EXTRA_SESSION_ID));
         assertEquals("authorization", intent.getStringExtra(DropInClient.EXTRA_AUTHORIZATION));
 
-        DropInRequest dropInRequestExtra = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST);
+        Bundle bundle = intent.getParcelableExtra(DropInClient.EXTRA_CHECKOUT_REQUEST_BUNDLE);
+        bundle.setClassLoader(DropInRequest.class.getClassLoader());
+        DropInRequest dropInRequestExtra = bundle.getParcelable(DropInClient.EXTRA_CHECKOUT_REQUEST);
         assertTrue(dropInRequestExtra.isVaultManagerEnabled());
     }
 

